### PR TITLE
lightning: job range keys and region split keys are 2 concepts (part 2)

### LINKED ISSUE #55374: Global sort IMPORT INTO stuck after region size is larger than 96MB
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1、tidb_enable_dist_task='on' and enable global sort
2、run import into

### 2. What did you expect to see? (Required)
import into can success

### 3. What did you see instead (Required)
import into blocked 

img_v3_02dn_fa67a709-910a-4240-8090-d52d4496f07g

the status of import job is not finished after 2h0m0s (now: 2024-08-12 23:45:55, jobId: 60004, step: importing, status: running)
operatorLogs:
[2024-08-12 21:45:54] ###### start import into
import into user_data1 from 's3://qe-testing/brie/lightning-mhy-100G-csv/sysbench.user_data1.000001*.csv?access-key=xxxxxx&secret-access-key=xxxxxx&endpoint=http%3a%2f%2fks3-cn-beijing-internal.ksyuncs.com&force-path-style=false&region=Beijing&provider=ks' WITH DETACHED,thread=16,skip_rows=1
[2024-08-12 21:45:54] ###### wait for import job to finish
[2024-08-12 23:45:55] ###### wait for import job to finish timeout (2h0m0s)
select id, table_schema, table_name, step, status,parameters from mysql.tidb_import_jobs where start_time >= '2024-08-12 21:45:54' and table_schema = 'importtest1' and table_name = 'user_data1'
jobId: 60004, table_schema: importtest1, table_name: user_data1, step: importing, status: running

### 4. What is your TiDB version? (Required)
./tidb-server -V
 Release Version: v8.3.0-alpha
Edition: Community
Git Commit Hash: 6df7aba3b84cd6a6ff54f5a10336b242fee3584d
Git Branch: heads/refs/tags/v8.3.0-alpha
UTC Build Time: 2024-08-10 11:52:15
GoVersion: go1.21.10
Race Enabled: false
Check Table Before Drop: false
Store: unistore
2024-08-12T21:45:47.369+0800

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: close #55374 

Problem Summary:

### What changed and how does it work?

1. common.Engine interface has changed. Now it doesn't have a region split size parameter. Implementations can choose split keys or range keys when constructed, or set the region split size parameter as a member.
2. LoadIngestData now returns multiple ranges instead of one range. These ranges should be considered with actual regions in cluster and convereted to regionJob
3. LoadIngestData now return data and ranges based on implementation's internal information rather than arguments. For example, for external engine it will use jobKeys.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: pkg/lightning/backend/external/engine.go [hunk 1]
```diff
@@ -350,16 +383,16 @@ func (e *Engine) loadBatchRegionData(ctx context.Context, startKey, endKey []byt
...
 ) error {
 	// try to make every worker busy for each batch
 	regionBatchSize := e.workerConcurrency
 	failpoint.Inject("LoadIngestDataBatchSize", func(val failpoint.Value) {
 		regionBatchSize = val.(int)
 	})
-	for i := 0; i < len(regionRanges); i += regionBatchSize {
-		err := e.loadBatchRegionData(ctx, regionRanges[i].Start, regionRanges[min(i+regionBatchSize, len(regionRanges))-1].End, outCh)
+	for start := 0; start < len(e.jobKeys)-1; start += regionBatchSize {
+		end := min(1+start+regionBatchSize, len(e.jobKeys))
```
[the remark this answers is not in the document]
Comment (lance6716) [reply pkg/lightning/backend/external/engine.go:395]: https://github.com/pingcap/tidb/pull/55759/commits/afb5e36fc438d594f346e3cefc4d99a4edb274af

File: pkg/lightning/backend/local/engine.go [hunk 2]
```diff
@@ -1084,15 +1101,33 @@ func (e *Engine) Finish(totalBytes, totalCount int64) {
...
-	regionRanges []common.Range,
-	outCh chan<- common.DataAndRange,
-) error {
-	for _, r := range regionRanges {
+	outCh chan<- common.DataAndRanges,
+) (err error) {
+	jobRangeKeys := e.regionSplitKeysCache
+	// when the region is large, we need to split to smaller job ranges to increase
+	// the concurrency.
+	if jobRangeKeys == nil || e.regionSplitSize > 2*int64(config.SplitRegionSize) {
```
[the remark this answers is not in the document]
Comment (lance6716) [reply pkg/lightning/backend/local/engine.go:1109]: This code is only run for local engine which stores data on disk, it has less memory usage. And this adjustment is moved from existing codes where you also commented 

```
	if localEngine, ok := engine.(*Engine); ok {
		// when use dynamic region feature, the region may be very big, we need
```

File: pkg/lightning/backend/local/local.go [hunk 3]
```diff
@@ -942,34 +938,13 @@ func (local *Backend) prepareAndSendJob(
...
 	engine common.Engine,
-	jobRanges []common.Range,
 	regionSplitSize, regionSplitKeys int64,
 	jobToWorkerCh chan<- *regionJob,
 	jobWg *sync.WaitGroup,
 ) error {
-	logger := log.FromContext(ctx)
-	// for external engine, it will split into smaller data inside LoadIngestData
-	if localEngine, ok := engine.(*Engine); ok {
-		// when use dynamic region feature, the region may be very big, we need
```
[the remark this answers is not in the document]
Comment (lance6716) [reply pkg/lightning/backend/local/local.go:953]: This adjustment is moved into local engine's LoadIngestData

**Review by D3Hunter — APPROVED** [review]

File: pkg/ddl/backfilling_import_cloud.go [hunk 4]
```diff
@@ -112,6 +112,7 @@ func (m *cloudImportExecutor) RunSubtask(ctx context.Context, subtask *proto.Sub
 	// compatible with old version task meta
 	jobKeys := sm.RangeJobKeys
 	if jobKeys == nil {
+		// TODO(lance6716): we have changed the encoding requirement, can it really be compatible?
```
Comment (lance6716) [inline pkg/ddl/backfilling_import_cloud.go:115]:
```suggestion
```

File: pkg/lightning/backend/external/engine.go [hunk 5]
```diff
@@ -327,21 +331,50 @@ func (e *Engine) loadBatchRegionData(ctx context.Context, startKey, endKey []byt
...
+			return err3
 		}
-		return nil
+		ranges = append(ranges, common.Range{
+			Start: prev,
+			End:   cur,
+		})
+		prev = cur
+	}
+	// last range key may be a nextKey so we should try to remove the trailing 0 before decode
```
Comment (lance6716) [inline pkg/lightning/backend/external/engine.go:350]:
```suggestion
	// last range key may be a nextKey so we should try to remove the trailing 0 if decoding failed
```

File: pkg/lightning/backend/local/region_job.go [hunk 6]
```diff
@@ -138,6 +139,113 @@ type injectedIngestBehaviour struct {
...
+		curRegion      = sortedRegions[curRegionIdx].Region
+		curRegionStart []byte
+		curRegionEnd   []byte
+	)
+
+	_, curRegionStart, _ = codec.DecodeBytes(curRegion.StartKey, nil)
+	_, curRegionEnd, _ = codec.DecodeBytes(curRegion.EndKey, nil)
+
+	for _, jobRange := range sortedJobRanges {
+		// build the job and move to next region for these the cases:
```
Comment (lance6716) [inline pkg/lightning/backend/local/region_job.go:200]:
```suggestion
		// build the job and move to next region for these cases:
```

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Review by yudongusa — APPROVED** [review]
LGTM

**Pull request merged**